DM-36111: miscellaneous improvements to registry support classes.

TallJimbo opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

TallJimbo committed:
Rework handling of governor dimension restrictions.

It took me a while to figure out why the set-like
GovernorDimensionRestriction class kept needing new set-op method
variants for ~every context in which it was being used, but the answer
is that it was really a mapping of sets (one for each governor
dimension), and hence when instances have different dimensions
(mapping keys) dimensions, their relationship is not set-like.

The best solution seems to be to remove that class in favor of the
kind of mapping-of-sets that backed it, and move its convenience
functionality that's important for CollectionSummary to
CollectionSummary itself.

I changed some attribute names and the module name for collecton
summary to get them consistent with now-preferred style; it made sense
to do this at the same time because I wanted to find all code using
CollectionSummary in order to make sure it was updated appropriately,
and renames made MyPy very good at that.  I also switched from
Dimension instance keys to `str` dimension name keys, since I think
we'll be doing more of that with RFC-834 adopted.

Finally, I've changed the logic in the user-expression checking code
to track all dimensions, and then filter down to just the governor
dimensions at the boundary with the rest of the query system.  This
just seemed the best for overall code simplicity.

TallJimbo committed:
Allow timespan selects/constraints from non-CALIBRATION collections.

The timespan for a dataset subquery for a non-CALIBRATION collection
is hereby defined to be a fully unbounded timespan, which will let us
UNION those subqueries against CALIBRATION-collection queries for the
same dataset before doing a temporal join against a dimension like
exposure or visit.

It would make sense to change the behavior of queryDatasetAssociations
to associate non-CALIBRATION collections with a full timespan instead
of None at the same time, but that's an API change that I don't want
to tackle right now.  Eventually I'd like to replace
queryDatasetAssocations entirely with a method-chaining interface to
get associations from queryDatasets, and I think that RFC wil be the
time to tackle this.

TallJimbo committed:
Add __eq__, __hash__, and __repr__ to CollectionRecord.

TallJimbo committed:
Modify how collection rank is computed in dataset subqueries.

This case-based approach is much closer to Andy Salnikov's initial
implementation of find-first dataset queries; I think it's likely that
changing it was at most unnecessary performance-wise (I was looking
for ways to simplify a query the planner was having trouble with, but
this construct - despite being hard for humans to parse when reading
the SQL - probably wasn't playing a role in that).  And some other
query system changes have made it preferable from a code-organization
standpoint, providing it's no worse from a performance standpoint.

To that end, in some quick benchmarks (on PostgreSQL, querying 16
collections with the same dataset types and a spatial constraint to
make the overall query nontrivial), this was consistently faster than
the old, UNION-based rank expression, but only barely; I'm by no means
certain that difference is significant, but I'm at least pretty
confident that this is no worse for performance.

TallJimbo committed:
Add Database interface for using constant rows in queries.

TallJimbo committed:
Add missing copyright header.

TallJimbo committed:
Add temporary table context manager to Session.

This is a very thin wrapper around existing methods to create and drop
temporary tables, but it's a convenient one, especially when using
contextlib.ExitStack.

TallJimbo committed:
Rework timespan SQL expression classes.

This removes the analogous expression class for spatial regions, as
well as the base class they shared.  We need an abstraction layer
above SQLAlchemy's for timespans, since those can have a
representation that involves multiple columns in some databases, and
SQLAlchemy's ColumnElement can't handle that.  But our regions are
always just an opaque blob column, so that abstraction layer wasn't
carrying its weight, despite the nice symmetry between regions and
timespans in other respects.  A new classmethod constructor has been
added to FieldSpec to take care of the only important thing this class
did - holding the defaults for how region columns should be defined.

That results in a number of abstract methods now being first defined
in TimespanDatabaseRepresentation.  For the most part, these are just
transfers from the old base class with slightly adjusted docs, but
there are a few small updates:

- fromLiteral now accepts None as a way to support NULL expressions;

- fromSelectable has been replaced from_columns, since SQLAlchemy is
  moving towards being much more strict about the difference between
  the columns of a SELECT statement and the columns of a FromClause,
  such as a table or an aliased subquery, and in the future we won't
  be able to access a ".columns" across the board (e.g.  on Select
  objects, it'll be "exported_columns" instead).

- we now use tuples instead of generators for a few methods, since
  generators for 1- or 2-element results involves a lot more overhead
  with no real gain.

TallJimbo committed:
Support region intersections in DataCoordinate.

This is enabled by new sphgeom functionality added on DM-34721.

TallJimbo committed:
Remove superfluous region arguments in Query methods.

These allowed the query's internal region to be overridden, but they
were a solution in search of a problem.

TallJimbo committed:
Export Database's Session helper class symbol.

TallJimbo committed:
Make DatasetAssociation immutable and hashable.

TallJimbo committed:
Handle single-time arguments in Timespan.overlaps.

(and TimespanDatabaseRepresentation.overlaps).

This makes the Python interface more consistent with the string
expression language, which uses OVERLAPS for timespan-time arguments,
and it will sidestep a problem with function name resolution in the new
daf_relation expression system, since normally that looks in the
operator module first and a method second, and "contains" (but not
"overlaps") is a member of the operator module.

TallJimbo committed:
Allow custom SQLAlchemy types to be initialized with no size.

This is useful for adding types to literals, which don't need to be
sized the same way that columns in tables do.

TallJimbo committed:
Add quoting to SQLite string-column check constraints.

This was previously blocking column names with non-standard SQL
characters.

TallJimbo committed:
Ensure Registry exception tests actually try to execute a query.

TallJimbo committed:
Forward correct region from QueryBuilder to Query.

andy-slac commented on python/lsst/daf/butler/registry/queries/_structs.py:
```diff
@@ -223,10 +224,10 @@ class QueryWhereClause:
-    imposed by the string expression or data ID
-    (`GovernorDimensionRestriction`).
+    imposed by the string expression and/or data ID
+    (`Mapping` [ ``set`,  `~collections.abc.Set` ] ]).
```
Double backtick, and `set` should be `str`?

andy-slac commented on python/lsst/daf/butler/registry/datasets/byDimensions/summaries.py:
```diff
@@ -294,20 +291,16 @@ def refresh(self, get_dataset_type: Callable[[int], DatasetType]) -> None:
             # None/NULL (many will be in general, because these enter the query
             # via LEFT OUTER JOIN).
+            summary.dataset_types.add(datasetType)
             for dimension in self._tables.dimensions:
                 value = row[dimension.name]
                 if value is not None:
-                    summary.dimensions.add(dimension, value)
+                    summary.governors.setdefault(dimension.name, set()).add(value)
```
Should be this using `summary.add_data_ids()` instead of modifying data members directly?

TallJimbo replied:
No, I left the data member as public (with a public annotation as a mutable type) because I wasn't bothered by external code doing its own updates to it, and trying to make this block produce data IDs just so they can be unpacked into a different structure didn't seem worthwhile; I prefer the transparency of seeing exactly what's going on over encapsulation in this case.

andy-slac commented on python/lsst/daf/butler/registry/queries/expressions/check.py:
```diff
@@ -336,33 +333,33 @@ def visitInner(self, branches: Sequence[TreeSummary], form: NormalForm) -> Inner
-                governor = branch.dataIdKey
-                value = summary.governors.setdefault(governor, branch.dataIdValue)
-                if value != branch.dataIdValue:
+            # Test whether this branch has a form like '<dimension>=<value' (or
```
Angle bracket missing after `<value`.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -1724,6 +1725,67 @@ def query(
+
+        Notes
+        -----
+        This is used by `Session.upload` to decide when `constant_rows` can be
```
I don't see `Session.upload`, is it in one of your other branches?

TallJimbo replied:
It's something I ended up removing, because I found that calling code always wanted to manually inspect this value to decide how whether to make a temporary table or use `constant_rows` itself (since one involves a resource whose lifetime should be managed, and the other does not).  I've dropped this first sentence from the docs and reworded the next accordingly.

andy-slac commented on python/lsst/daf/butler/core/timespan.py:
```diff
@@ -647,6 +719,110 @@ def fromLiteral(cls: Type[_S], timespan: Timespan) -> _S:
+
+        Returns
+        -------
+        timespan
```
Add `: Timespan, optional`?

TallJimbo replied:
Actually want it to be "`Timespan` or `None`" in this case, but yes.

andy-slac commented on python/lsst/daf/butler/registry/datasets/byDimensions/_storage.py:
```diff
@@ -396,6 +397,14 @@ def select(
             # create a dummy subquery that we know will fail.
             tags_query = SimpleQuery()
             tags_query.join(self._tags, **kwargs)
+            # If the timespan is requested or contrained, simulate a
```
typo: contrained

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -265,6 +265,31 @@ def dropTemporaryTable(self, table: sqlalchemy.schema.Table) -> None:
             raise TypeError(f"Table {table.key} was not created by makeTemporaryTable.")
 
+    @contextmanager
+    def temporary_table(self, spec: ddl.TableSpec, name: Optional[str]) -> Iterator[sqlalchemy.schema.Table]:
```
I don't see this used anywhere, I guess it's also for one of the other branches?

TallJimbo replied:
Correct.  An upcoming branch will use it.  I suppose I could also use it to simplify the existing `[Something]QueryResult.materialize` context managers, but I'll be replacing them completely in one of those upcoming branches, too.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -263,6 +265,31 @@ def dropTemporaryTable(self, table: sqlalchemy.schema.Table) -> None:
+        try:
+            yield table
+        finally:
+            self.dropTemporaryTable(table)
```
I think Coverity tries to say that there is no unit test for the whole `temporary_table` but somehow it only points to a single line.

andy-slac reviewed (approved):
Looks good, few minor comments.

TallJimbo committed:
Add changelog entries for publicly-visible changes.

This branch also includes a number of behind-the-scenes changes that
I don't think merit a changelog entry.

TallJimbo committed:
Fix variable-shadowing bug in dimension export kwarg-handling.

TallJimbo committed:
Fix doc typo.

TallJimbo committed:
Fix doc return-type references.

TallJimbo committed:
Add SQL cast to fix from-SQL conversion of timespan literals.

TallJimbo committed:
Remove workaround for SQLAlchemy bug that has been fixed.

We already seem to require SQLAlchemy 1.4, which includes the fix we
submitted upstream.

TallJimbo commented:
@andy-slac , I'm sending this back to you for another look because Jenkins revealed a problem that led to a non-trivial change, and then I decided to throw in a couple more miscellaneous fixes.  In particular:

- 465ebeb8572c0e7a7032d5917f50d2a6198fc47a has been modified to define the timespan for a non-CALIBRATION collection to a fully-unbounded one, instead of NULL.  That's logically the behavior we want in the query system anyway, and it avoids the need for special IS NULL handling in queries to address that downstream problem Jenkins discovered.

- That change led to d6df482bddf77552db46b7f58ae49aed206bf664, fixing a preexisting bug in how SQLAlchemy converts literals in queries (see also https://github.com/sqlalchemy/sqlalchemy/discussions/8540).

- 5aa83fa7ef2a04e75e2befb7a682dc4251d7f61d is a fix for a bug reported on Slack a couple of weeks ago that I was going to put on this branch but forgot about as I disappeared for vacation.

- 9cb3aec1111283e95904b80b741d7efe09345fb0, 38c22a6c622a56693adabbb323e8fb47ec593258, and 3909cdf3ea2cc199cc197e6f03f6993df814aedf are tiny fixed for things I noticed while I was working through review comments.

andy-slac reviewed (approved):
Looks good, no new comments.
